### PR TITLE
RD-2481 Added "Upload all plugins" button

### PR DESCRIPTION
**Pull request by kubama**
The button was added as a standard data table action button:
image

Some implementation notes:
- plugin upload state was moved to redux - this way it can be tracked application wide, the user can switch between pages and the state is not lost
- when plugin upload is completed an event is emitted - as above, when user switches between pages and comes back to the catalog widget he can still see notifications, even if the upload was started in a different widget instance
- plugins list was migrated to functional component, as it makes working with events and redux easier

### CODE REVIEW

**Comment by qooban** [comment]
I'll review this PR on Monday.

**Comment by stepek** [comment]
will review on Monday

**Comment by kubama** [comment]
System tests: https://jenkins.cloudify.co/view/UI/job/Stage-UI-System-Test/1075/

**Comment by kubama** [comment]
There is one system test failure, but the same test fails on `master`, so I consider it a pass.

File: widgets/pluginsCatalog/src/PluginsCatalogList.tsx [hunk 1]
```diff
@@ -1,7 +1,9 @@
-import type { ComponentProps, ComponentType } from 'react';
+import type { ComponentProps, ComponentType, FunctionComponent } from 'react';
+import { useState } from 'react';
 import styled from 'styled-components';
-import { compact, isEmpty, isEqual, find, map, without } from 'lodash';
+import { compact, find, isEmpty, map, without } from 'lodash';
 
+import { useDispatch, useSelector } from 'react-redux';
```
Comment (qooban) [inline widgets/pluginsCatalog/src/PluginsCatalogList.tsx:6]: Hmm... Shouldn't we use `ReactRedux.useDispatch` and `ReactRedux.useSelector` in `widgets`?

File: widgets/pluginsCatalog/src/widget.tsx [hunk 2]
```diff
@@ -56,6 +57,10 @@ Stage.defineWidget<unknown, PluginsCatalogResponse | Error, PluginsCatalogWidget
             formattedData = _.sortBy(data, item => item.pluginDescription.title);
         }
 
-        return <PluginsCatalogList widget={widget} items={formattedData} toolbox={toolbox} />;
+        return (
+            <Provider store={toolbox.getStore()}>
+                <PluginsCatalogList widget={widget} items={formattedData} toolbox={toolbox} />
+            </Provider>
```
Comment (qooban) [inline widgets/pluginsCatalog/src/widget.tsx:63]: Is that necessary? Recently I used redux store in "Number of Deployments" widget and did not encounter that need. We also use redux store in "Deployments View" widget.

File: app/utils/Toolbox.ts [hunk 3]
```diff
@@ -73,6 +73,10 @@ class Toolbox implements Stage.Types.Toolbox {
         this.store.dispatch((selectPageByName(pageName, context) as unknown) as AnyAction);
     };
 
+    getStore() {
+        return this.store;
+    }
+
```
Comment (qooban) [inline app/utils/Toolbox.ts:79]: I'm not sure we need it. See my other two comments about using Redux store in widgets.

**Review by qooban — COMMENTED** [review]
I did not review `PluginsCatalogList.tsx` and `plugins_catalog_spec.ts` yet. 

I've added 3 comments. All of them are related to the same thing - how to handle redux store in widgets. I'll continue the review once we finalize that topic.

**Comment by kubama** [comment]
> I did not review `PluginsCatalogList.tsx` and `plugins_catalog_spec.ts` yet.
> 
> I've added 3 comments. All of them are related to the same thing - how to handle redux store in widgets. I'll continue the review once we finalize that topic.

I addressed the comments - implementation is now aligned to `ReactRedux`

**Review by stepek — DISMISSED** [review]

File: test/cypress/integration/widgets/plugins_catalog_spec.ts [hunk 4]
```diff
@@ -106,4 +106,26 @@ describe('Plugins Catalog widget', () => {
...
+    it('should upload all plugins', () => {
+        // eslint-disable-next-line security/detect-non-literal-regexp
+        cy.intercept('POST', new RegExp(`console/plugins/upload.*title=AWS`)).as('awsPluginUpload');
+        cy.contains('Upload all plugins').click().should('be.disabled');
+        cy.get('.pluginsCatalogWidget tr button').should('be.disabled');
+        // eslint-disable-next-line security/detect-non-literal-regexp
+        cy.intercept('POST', new RegExp(`console/plugins/upload.*title=Ansible`)).as('ansiblePluginUpload');
+        cy.wait('@awsPluginUpload', { responseTimeout: uploadPluginTimeout });
+        cy.contains('AWS successfully uploaded');
+        cy.contains('.pluginsCatalogWidget tr', 'AWS').find('button.loading').should('not.exist');
```
Comment (qooban) [inline test/cypress/integration/widgets/plugins_catalog_spec.ts:119]: Optional:

Maybe for easier reading we can have:
1. Both intercepts at the top
2. Some spacing between setup, test and assertion blocks
WDYT?

```suggestion
        // eslint-disable-next-line security/detect-non-literal-regexp
        cy.intercept('POST', new RegExp(`console/plugins/upload.*title=AWS`)).as('awsPluginUpload');
        // eslint-disable-next-line security/detect-non-literal-regexp
        cy.intercept('POST', new RegExp(`console/plugins/upload.*title=Ansible`)).as('ansiblePluginUpload');
        
        cy.contains('Upload all plugins').click().should('be.disabled');
        
        cy.get('.pluginsCatalogWidget tr button').should('be.disabled');
        cy.wait('@awsPluginUpload', { responseTimeout: uploadPluginTimeout });
        cy.contains('AWS successfully uploaded');
        cy.contains('.pluginsCatalogWidget tr', 'AWS').find('button.loading').should('not.exist');
```

**Review by qooban — APPROVED** [review]
Thanks for addressing my comments.

I think importing Redux was the main reason for widget bundle size increase as right now it dropped to:
```
[2021-08-16T07:43:28.988Z]   Widgets
[2021-08-16T07:43:28.988Z]   Size limit:   3.13 MB
[2021-08-16T07:43:28.988Z]   Size:         3.1 MB
```

File: test/cypress/integration/widgets/plugins_catalog_spec.ts (hunk 4 above)
Comment (kubama) [reply test/cypress/integration/widgets/plugins_catalog_spec.ts]: Actually my intension here was to test that the second plugin upload starts after the first one, not in parallel to the first one. Moving the second intercept to the top would make the test to pass regardless of requests timing. I consider it important as launching all plugin uploads at once was causing the backend to freeze for a while making the application unresponsive.

I added empty lines as suggested though. Also added a comment.

Comment (qooban) [reply test/cypress/integration/widgets/plugins_catalog_spec.ts]: OK, got it, sounds reasonable to have it in that order then. Thanks for adding the comment about it.

**Review by qooban — APPROVED** [review]

**Pull request merged**